src/Coq.hs

```diff
@@ -0,0 +1,410 @@
+      [ coqprop <$> conds
+      , [retname <> " " <> stateVar <> " " <> arguments i <> " " <> retexp r]
+      ]
+
+retVal _ = return ""
+
+-- | produce a state value from a list of storage updates
+-- 'handler' defines what to do in cases where a given name isn't updated
+stateval
+  :: Store
+  -> (Id -> SlotType -> T.Text)
+  -> [StorageUpdate]
+  -> T.Text
+stateval store handler updates =
+
+  stateConstructor <> " " <> T.intercalate " "
+    (map (valuefor updates) (M.toList store))
+
+  where
+
+  valuefor :: [StorageUpdate] -> (Id, SlotType) -> T.Text
+  valuefor updates' (name, t) =
+    case find (f name) updates' of
+      Nothing -> parens $ handler name t
+      Just (IntUpdate (DirectInt _ _) e) -> parens $ coqexp e
+      Just (IntUpdate (MappedInt _ name' args) e) -> lambda (NE.toList args) 0 e name'
+      Just (BoolUpdate (DirectBool _ _) e)  -> parens $ coqexp e
+      Just (BoolUpdate (MappedBool _ name' args) e) -> lambda (NE.toList args) 0 e name'
+      Just (BytesUpdate _ _) -> error "bytestrings not supported"
+
+  -- filter by name
+  f n (IntUpdate (DirectInt _ n') _)
+    | n == n' = True
+  f n (IntUpdate (MappedInt _ n' _) _)
+    | n == n' = True
+  f n (BoolUpdate (DirectBool _ n') _)
+    | n == n' = True
+  f n (BoolUpdate (MappedBool _ n' _) _)
+    | n == n' = True
+  f _ _ = False
+
+  -- represent mapping update with anonymous function
+  lambda :: [ReturnExp] -> Int -> Exp a -> Id -> T.Text
+  lambda [] _ e _ = parens $ coqexp e
+  lambda (x:xs) n e m = let name = anon <> T.pack (show n) in parens $ "fun "
+    <> name
+    <> " => if "
+    <> name <> eqsym x <> retexp x <> " then " <> lambda xs (n + 1) e m <> " else "
+    <> T.pack m <> " " <> stateVar <> " " <> lambdaArgs n
+
+  lambdaArgs n = T.intercalate " " $ map (\x -> anon <> T.pack (show x)) [0..n]
+
+  eqsym (ExpInt _) = " =? "
+  eqsym (ExpBool _) = " =?? "
+  eqsym (ExpBytes _) = error "bytestrings not supported"
+
+-- | produce a block of declarations from an interface
+interface :: Interface -> T.Text
+interface (Interface _ decls) =
+  T.intercalate " " (map decl decls) where
+  decl (Decl t name) = parens $ T.pack name <> " : " <> abiType t
+
+arguments :: Interface -> T.Text
+arguments (Interface _ decls) =
+  T.intercalate " " (map (\(Decl _ name) -> T.pack name) decls)
+
+-- | coq syntax for a slot type
+slotType :: SlotType -> T.Text
+slotType (StorageMapping xs t) =
+  T.intercalate " -> " (map abiType (NE.toList xs ++ [t]))
+slotType (StorageValue abitype) = abiType abitype
+
+-- | coq syntax for an abi type
+abiType :: AbiType -> T.Text
+abiType (AbiUIntType _) = "Z"
+abiType (AbiIntType _) = "Z"
+abiType AbiAddressType = "address"
+abiType AbiStringType = strMod <> ".string"
+abiType a = error $ show a
+
+-- | coq syntax for a return type
+returnType :: ReturnExp -> T.Text
+returnType (ExpInt _) = "Z"
+returnType (ExpBool _) = "bool"
+returnType (ExpBytes _) = "bytestrings not supported"
+
+-- | default value for a given type
```

Comment:
actually, all state variables must be defined and given an initial value in the constructor right now, so this seems superfluous

Reply:
it's not superfluous because of mappings. `mapping(t => t') m := []` doesn't have a sensible meaning if there isn't a default value for the type `t'`.

Reply:
Right, ok, this is good